[LoopVectorize] Add support for reverse loops in isDereferenceableAndAlignedInLoop

david-arm opened the pull request:
Currently when we encounter a negative step in the induction
variable isDereferenceableAndAlignedInLoop bails out because
the element size is signed greater than the step. This patch
adds support for negative steps in cases where we detect the
start address for the load is of the form base + offset. In
this case the address decrements in each iteration so we need
to calculate the access size differently. I have done this by
caling getStartAndEndForAccess from LoopAccessAnalysis.cpp.

The motivation for this patch comes from PR #88385 where a
reviewer requested reusing isDereferenceableAndAlignedInLoop,
but that PR itself does support reverse loops.

The changed test in LoopVectorize/X86/load-deref-pred.ll now
passes because previously we were calculating the total access
size incorrectly, whereas now it is 412 bytes and fits
perfectly into the alloca.

fhahn commented on llvm/lib/Analysis/Loads.cpp:
```diff
@@ -318,11 +321,24 @@ bool llvm::isDereferenceableAndAlignedInLoop(LoadInst *LI, Loop *L,
+        // We can safely use the new base because the decrementing pointer is
+        // always guaranteed to be >= new base. The total access size needs to
+        // take into account the start offset and the loaded element size.
+        AccessSize = Offset->getAPInt() + EltSize;
```
Does this need an overflow check and bail out in case we have a narrow IV where incrementing by EltSize wraps? Alternatively we could probably also perform the computation in  always in something like i64?

david-arm replied:
Good point! I've added the overflow check.

fhahn replied:
Thanks for the update! Do you think it would be possible to add a test case that covers the overflow case (e.g. with an i8 induction that wraps when adding the element size?

david-arm replied:
I tried doing that with an i8, but the offset return by SCEV is always 64-bit. So in practice I'd need a test with an alloca just less than UINT64_MAX bytes in size and an offset just less than UINT64_MAX, in order to exercise the overflow case. I wasn't sure if this was fragile, i.e. are allocas of such a large size even legal given that I don't think GEPs even support an offset greater than SINT64_MAX (since offsets are signed)? Having said that, I'm open to ideas about how to write a sensible test for this!

fhahn replied:
Would something like the below work? I might be missing something, but it looks like the overflow check should trigger there?


```

target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128-ni:1"

declare void @init(ptr nocapture nofree)


define void @test(ptr nocapture noundef writeonly %dest) {

entry:
  %local_dest = alloca [128 x i8], align 4
  %local_src = alloca [128 x i8], align 4
  %local_cmp = alloca [128 x i8], align 4
  call void @init(ptr %local_src)
  call void @init(ptr %local_cmp)
  br label %for.body

for.body:
  %iv = phi i8 [ 255, %entry ], [ %iv.next, %for.inc ]
  %arrayidx = getelementptr inbounds [1024 x i8], ptr %local_cmp, i8 0, i8 %iv
  %0 = load i8, ptr %arrayidx, align 1
  %cmp3.not = icmp eq i8 %0, 3
  br i1 %cmp3.not, label %for.inc, label %if.then

if.then:
  %arrayidx5 = getelementptr inbounds [1024 x i8], ptr %local_src, i8 0, i8 %iv
  %1 = load i8, ptr %arrayidx5, align 1
  %mul = shl nsw i8 %1, 2
  %arrayidx7 = getelementptr inbounds [1024 x i8], ptr %local_dest, i8 0, i8 %iv
  store i8 %mul, ptr %arrayidx7, align 1
  br label %for.inc

for.inc:
  %iv.next = add nsw i8 %iv, -1
  %cmp2.not = icmp eq i8 %iv, 200
  br i1 %cmp2.not, label %for.cond.cleanup, label %for.body

for.cond.cleanup:
  call void @llvm.memcpy.p0.p0.i8(ptr noundef nonnull align 4 dereferenceable(1024) %dest, ptr noundef nonnull align 4 dereferenceable(1024) %local_dest, i8 1024, i1 false)
  ret void
}
```

david-arm replied:
> Would something like the below work? I might be missing something, but it looks like the overflow check should trigger there?

Yep, that's exactly the test I tried and the offset produced was a 64-bit APInt. I'll try again with your IR just in case you've written it differently to me.

david-arm replied:
Right, so the offset is coming out as 64-bits, but it's probably not what you were expecting - Offset = 0xFFFFFFFFFFFFFFFF. Again, I think it's because the offset in a GEP is treated as being signed so (i8 255) is equivalent to (i8 -1) and is being sign-extended to (i64 -1). So yes technically when we try to (unsigned) add the element size (1 byte) to the offset it overflows, but I'm not sure we can rely upon that for testing overflow behaviour.

This actually makes me wonder whether there are already problems with the existing code, which seems to unconditionally treat the offset as unsigned. For example, should this code

```
       // TODO: generalize if a case found which warrants
       if (Offset->getAPInt().urem(Alignment.value()) != 0)
         return false;
```

be rewritten to

```
       // TODO: generalize if a case found which warrants
       if (Offset->getAPInt().srem(Alignment.value()) != 0)
         return false;
```

etc?

david-arm replied:
Or perhaps bail out entirely if the offset is negative?

david-arm replied:
I think it might makes sense for now to bail out for negative values - it doesn't cause any existing tests to fail. It would also make it almost impossible to write an overflow test because you'd need something like an offset of 0x7FFFFFFFFFFFFFFF and an element size of 0x7FFFFFFFFFFFFFFF

fhahn replied:
I think it makes sense to restrict this to non-negative values, thanks! It would probably make sense to split off the patch with the extra bail-out separately. Also adding @preames who added some of that code originally I think

david-arm replied:
So I originally thought about splitting off the change in a separate patch, but I couldn't see any way of testing it by itself because the overflow checks catch this by accident. If you want I can split it off into a separate PR - there just won't be any tests for it.

david-arm replied:
Created https://github.com/llvm/llvm-project/pull/99490 for the negative offset check.

david-arm commented:
* Rebase + fix tests after rebase.

preames commented:
An alternate approach here would be to compute the AccessSize as a SCEV (which should just be umax(End,Start)-umin(Start,End) + ElementSize, and then check to see if it's a SCEVConstant.  End should be evaluateAtIteration on the AddRec.  The base would then be umin(End, Start) as a SCEV, and then see if it's an understandable one.  

In general, doing the math manually seems like a recipe for error here.

david-arm commented:
> An alternate approach here would be to compute the AccessSize as a SCEV (which should just be umax(End,Start)-umin(Start,End) + ElementSize, and then check to see if it's a SCEVConstant. End should be evaluateAtIteration on the AddRec. The base would then be umin(End, Start) as a SCEV, and then see if it's an understandable one.
> 
> In general, doing the math manually seems like a recipe for error here.

Thanks for this suggestion @preames! It was a bit fiddly to get it working without using SCEVExpander, but I do think it's now less prone to error. As part of this patch I've also added some tests for loops with strided accesses for positive and negative steps too as we didn't seem to be testing that.

david-arm commented:
* I've rebased the patch after landing the negative offset test. I also had to fix up a new test added recently.

david-arm commented:
@preames @fhahn After rebasing my early exit auto-vectorisation patch on top of this one, I realised some of the tests no longer vectorise because isDereferenceAndAlignedInLoop now returns false for some cases. In particular, if a function argument has the attribute `dereferenceable(100)` then the SCEV min/max code fails to return a SCEVConstant. This is because it is unable to determine that a calculation such as (ptr + offset) will not wrap. I don't think there is a fundamental issue blocking this - it's just that the function `ScalarEvolution::getRangeRef` knows about allocas, but doesn't know about the dereferenceable attribute.

So the question is - are we happy with accepting these minor regressions when using SCEV for now? Or is it worth creating a new patch to teach `ScalarEvolution::getRangeRef` new tricks first?

fhahn commented on llvm/lib/Analysis/Loads.cpp:
```diff
@@ -293,26 +293,65 @@ bool llvm::isDereferenceableAndAlignedInLoop(LoadInst *LI, Loop *L,
-    const auto *Offset = dyn_cast<SCEVConstant>(StartS->getOperand(0));
-    const auto *NewBase = dyn_cast<SCEVUnknown>(StartS->getOperand(1));
-    if (StartS->getNumOperands() == 2 && Offset && NewBase) {
+    const SCEV *End = AddRec->evaluateAtIteration(
```
Could we re-use the code for computing start and end from https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/LoopAccessAnalysis.cpp#L206

david-arm replied:
I think that's possible, but it's not a drop-in replacement and is quite inefficient in the current context. For example, `getStartAndEndForAccess` recalculates the trip count and element size which we've already calculated, although I could rip out a lot of the code in `isDereferenceableAndAlignedInLoop` so that I avoid doing this. The function also requires passing in a `PointerBounds` DenseMap, which we don't care about. I could probably refactor `getStartAndEndForAccess` by adding a new variant that doesn't require a map. Ultimately I will still need to extract an actual base `Value*` pointer from a SCEV, but perhaps this would reduce the overall complexity.

david-arm replied:
OK, I've done some refactoring and taught isDereferenceableAndAlignedInLoop to call getStartAndEndForAccess. This at least reduces some of the complexity and is actually an improvement on the existing code.

fhahn commented on llvm/lib/Analysis/Loads.cpp:
```diff
@@ -293,26 +293,65 @@ bool llvm::isDereferenceableAndAlignedInLoop(LoadInst *LI, Loop *L,
+      return false;
+
+    // Now calculate the total access size, which is (max - min) + element_size.
+    const SCEV *Diff = SE.getMinusSCEV(Max, Min);
```
Is the separate handling for SCEVUnknown and SCEVAddExpr still needed when expanding the bounds here? I'd expect that computing start/end of the add rec would work for both cases without the need for a distinction?

david-arm replied:
Yes I think so. That's because the Min value is a `SCEV*` whereas we need a `Value*` and you can only do that by analysing the SCEV type and handling it on a case-by-case basis. We cannot use the SCEVExpander here because that would create new IR. This is the downside to using SCEV here.

fhahn commented on llvm/lib/Analysis/Loads.cpp:
```diff
@@ -293,26 +293,65 @@ bool llvm::isDereferenceableAndAlignedInLoop(LoadInst *LI, Loop *L,
+    if (!ASC)
+      return false;
+
+    if (const SCEVUnknown *NewBase = dyn_cast<SCEVUnknown>(Min)) {
```
Retrieving the new base and access size here seems to unfortunately add a bit of complexity.

What could make this much simpler would be if we would get the base pointer (underlying object) and the # of known dereferenceable bits. Then construct SCEVs for the start and end of the object and check if the Start/End of the accessed pointer is >= and <= the bounds of the object.

https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/Loads.cpp#L97 would highlight a way to get the number of dereferenceable bytes

david-arm replied:
We could do certainly, but then we're effectively doing some of the work that's already done in `isDereferenceableAndAlignedPointer`. It's almost like we want to call `isDereferenceableAndAlignedPointer` on both the min and max addressed pointers in the loop.

david-arm replied:
Actually I still think we end up with a similar problem in that we need to discover the base `Value*` in order to call `getPointerDereferenceableBytes`. I understand what you're suggesting, but not sure it will reduce the complexity in practice. I'll have a think about it some more.

fhahn reviewed (commented):
> @preames @fhahn After rebasing my early exit auto-vectorisation patch on top of this one, I realised some of the tests no longer vectorise because isDereferenceAndAlignedInLoop now returns false for some cases. In particular, if a function argument has the attribute `dereferenceable(100)` then the SCEV min/max code fails to return a SCEVConstant. This is because it is unable to determine that a calculation such as (ptr + offset) will not wrap. I don't think there is a fundamental issue blocking this - it's just that the function `ScalarEvolution::getRangeRef` knows about allocas, but doesn't know about the dereferenceable attribute.
> 
> So the question is - are we happy with accepting these minor regressions when using SCEV for now? Or is it worth creating a new patch to teach `ScalarEvolution::getRangeRef` new tricks first?

That's interesting! Is it possible that the current tests don't catch this issue because `load-deref-pred-align.ll` doesn't have a variant with `dereferenceable()`?

david-arm commented:
> That's interesting! Is it possible that the current tests don't catch this issue because `load-deref-pred-align.ll` doesn't have a variant with `dereferenceable()`?

Yeah that's right, whereas my early exit autovec tests have a test for dereferenceable pointers. It turns out there is a relatively simple fix by teaching `ScalarEvolution::getRangeRef` to use `Value::getPointerDereferenceableBytes`. If I get time before I go away on holiday I'll get a patch up for this!

david-arm commented:
Rebased.

fhahn commented on llvm/test/Transforms/LoopVectorize/SystemZ/scalar-steps-with-users-demanding-all-lanes-and-first-lane-only.ll:
```diff
@@ -3,7 +3,7 @@
 target datalayout = "E-m:e-i1:8:16-i8:8:16-i64:64-f128:64-v128:64-a:8:16-n32:64"
 
-@src = external global [8 x i32], align 4
+declare void @init()
```
Why is this change needed?

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -203,29 +203,29 @@ RuntimeCheckingPtrGroup::RuntimeCheckingPtrGroup(
   if (SE->isLoopInvariant(PtrExpr, Lp)) {
     ScStart = ScEnd = PtrExpr;
   } else if (auto *AR = dyn_cast<SCEVAddRecExpr>(PtrExpr)) {
-    const SCEV *Ex = PSE.getSymbolicMaxBackedgeTakenCount();
```
as PSE caches the max symbolic BTC (I think) it's probably better to avoid the extra parameter.

fhahn commented on llvm/include/llvm/Analysis/LoopAccessAnalysis.h:
```diff
@@ -839,6 +839,12 @@ bool sortPtrAccesses(ArrayRef<Value *> VL, Type *ElemTy, const DataLayout &DL,
 bool isConsecutiveAccess(Value *A, Value *B, const DataLayout &DL,
                          ScalarEvolution &SE, bool CheckType = true);
 
+std::pair<const SCEV *, const SCEV *> getStartAndEndForAccess(
```
Could you add a comment here?

fhahn replied:
Could you move the comment from `llvm/lib/Analysis/LoopAccessAnalysis.cpp`, as it is more precise?

I think from the current description, it sounds like the returned value is the last address accessed in the loop, but unless I am missing something it is actually (address of last access)+access size.

david-arm replied:
Done!

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -203,29 +203,29 @@ RuntimeCheckingPtrGroup::RuntimeCheckingPtrGroup(
-static std::pair<const SCEV *, const SCEV *> getStartAndEndForAccess(
-    const Loop *Lp, const SCEV *PtrExpr, Type *AccessTy,
-    PredicatedScalarEvolution &PSE,
+std::pair<const SCEV *, const SCEV *> llvm::getStartAndEndForAccess(
```
might be good to add an overload without the extra EltSize argument for users here

david-arm replied:
Done

fhahn commented on llvm/test/Transforms/LoopVectorize/load-deref-pred-align.ll:
```diff
@@ -296,3 +296,432 @@ latch:
+  br label %for.body
+
+for.body:
+  %indvars.iv = phi i64 [ 511, %entry ], [ %indvars.iv.next, %for.inc ]
```
```suggestion
  %iv = phi i64 [ 511, %entry ], [ %indvars.iv.next, %for.inc ]
```

fhahn commented on llvm/test/Transforms/LoopVectorize/load-deref-pred-align.ll:
```diff
@@ -296,3 +296,432 @@ latch:
+  br i1 %cmp3.not, label %for.inc, label %if.then
+
+if.then:
+  %indvars.iv.strided = mul i64 %indvars.iv, 2
```
```suggestion
  %iv.strided = mul i64 %indvars.iv, 2
```

fhahn commented on llvm/test/Transforms/LoopVectorize/load-deref-pred-align.ll:
```diff
@@ -296,3 +296,432 @@ latch:
+  br label %for.body
+
+for.body:
+  %indvars.iv = phi i64 [ 1023, %entry ], [ %indvars.iv.next, %for.inc ]
```
```suggestion
  %iv = phi i64 [ 1023, %entry ], [ %indvars.iv.next, %for.inc ]
```

fhahn commented on llvm/test/Transforms/LoopVectorize/load-deref-pred-align.ll:
```diff
@@ -296,3 +296,432 @@ latch:
+  %cmp2.not = icmp eq i64 %indvars.iv, 0
+  br i1 %cmp2.not, label %for.cond.cleanup, label %for.body
+
+for.cond.cleanup:
```
```suggestion
exit:
```

fhahn commented on llvm/test/Transforms/LoopVectorize/load-deref-pred-align.ll:
```diff
@@ -296,3 +296,432 @@ latch:
+  br i1 %cmp2.not, label %for.cond.cleanup, label %for.body
+
+for.cond.cleanup:
+  call void @llvm.memcpy.p0.p0.i64(ptr noundef nonnull align 4 dereferenceable(1024) %dest, ptr noundef nonnull align 4 dereferenceable(1024) %local_dest, i64 1024, i1 false)
```
```suggestion
  call void @llvm.memcpy.p0.p0.i64(ptr %dest, ptr %local_dest, i64 1024, i1 false)
```

david-arm replied:
Done

fhahn commented on llvm/test/Transforms/LoopVectorize/load-deref-pred-align.ll:
```diff
@@ -296,3 +296,432 @@ latch:
+  br label %for.body
+
+for.body:
+  %indvars.iv = phi i64 [ 1023, %entry ], [ %indvars.iv.next, %for.inc ]
```
```suggestion
  %iv = phi i64 [ 1023, %entry ], [ %indvars.iv.next, %for.inc ]
```

fhahn commented on llvm/test/Transforms/LoopVectorize/load-deref-pred-align.ll:
```diff
@@ -296,3 +296,432 @@ latch:
+  br label %for.inc
+
+for.inc:
+  %indvars.iv.next = add nsw i64 %indvars.iv, -1
```
```suggestion
  %iv.next = add nsw i64 %indvars.iv, -1
```

david-arm replied:
Done

fhahn commented on llvm/lib/Analysis/Loads.cpp:
```diff
@@ -263,84 +264,88 @@ bool llvm::isDereferenceableAndAlignedInLoop(LoadInst *LI, Loop *L,
+  // Try to get the access size.
+  const SCEV *PtrDiff = SE.getMinusSCEV(Range.second, Range.first);
+  APInt MaxPtrDiff;
+  if (isa<SCEVConstant>(PtrDiff))
+    MaxPtrDiff = cast<SCEVConstant>(PtrDiff)->getAPInt();
+  else
+    MaxPtrDiff = SE.getUnsignedRangeMax(PtrDiff);
```
Sufficient to ? 
```suggestion
    MaxPtrDiff = SE.getUnsignedRangeMax(PtrDiff);
```

david-arm replied:
Done

fhahn commented on llvm/lib/Analysis/Loads.cpp:
```diff
@@ -263,84 +264,88 @@ bool llvm::isDereferenceableAndAlignedInLoop(LoadInst *LI, Loop *L,
+    return false;
+
+  const SCEV *EltSizeSCEV;
+  std::pair<const SCEV *, const SCEV *> Range = getStartAndEndForAccess(
```
```suggestion
  const auto &[AccessStart, AccessEnd] = getStartAndEndForAccess(
```

david-arm replied:
Done

david-arm committed:
[LoopVectorize] Add support for reverse loops in isDereferenceableAndAlignedInLoop

Currently when we encounter a negative step in the induction
variable isDereferenceableAndAlignedInLoop bails out because
the element size is signed greater than the step. This patch
adds support for negative steps in cases where we detect the
start address for the load is of the form base + offset. In
this case the address decrements in each iteration so we need
to calculate the access size differently. I have done this by
caling getStartAndEndForAccess from LoopAccessAnalysis.cpp.

The changed test in LoopVectorize/X86/load-deref-pred.ll now
passes because previously we were calculating the total access
size incorrectly, whereas now it is 412 bytes and fits
perfectly into the alloca.

david-arm commented:
Rebase and some minor refactoring to hopefully make it look a bit neater.

david-arm commented:
Rebase.

fhahn commented on llvm/lib/Analysis/Loads.cpp:
```diff
@@ -276,84 +277,85 @@ static bool AreEquivalentAddressValues(const Value *A, const Value *B) {
-  // For patterns with gaps (i.e. non unit stride), we are
-  // accessing EltSize bytes at every Step.
-  APInt AccessSize = TC * Step->getAPInt();
+  const SCEV *MaxBECount =
+      SE.getPredicatedSymbolicMaxBackedgeTakenCount(L, *Predicates);
+  if (isa<SCEVCouldNotCompute>(MaxBECount))
```
Extending to support predicated BTCs instead of constant trip counts should probably be done separately if possible?

It should probably also be possible add a test for predicated BTCs unless I am missing anything?

david-arm replied:
Yeah sorry, you're right. It shouldn't be part of this change. I think I just did it whilst addressing one of your review comments about reusing getStartAndEndForAccess. It should be fine to pass in `SE.getSmallConstantMaxTripCount` for now.

david-arm replied:
Actually, I can't pass in the result of `SE.getSmallConstantMaxTripCount` anymore because it returns an unsigned value. Instead I have changed it to call `SE.getPredicatedConstantMaxBackedgeTakenCount`, which is what `getSmallConstantMaxTripCount` does anyway.

fhahn replied:
Sounds good, thanks!

fhahn commented on llvm/lib/Analysis/Loads.cpp:
```diff
@@ -276,84 +277,85 @@ static bool AreEquivalentAddressValues(const Value *A, const Value *B) {
+  APInt MaxPtrDiff = SE.getUnsignedRangeMax(PtrDiff);
 
-  if (!Base)
+  // If the (max) pointer difference is > 32 bits then it's unlikely to be
+  // dereferenceable.
+  if (MaxPtrDiff.getActiveBits() > 32)
```
Is this needed to avoid some issues with bitwidths of expressions below or can this be dropped? Whether it is unlikely or not may depend on the info from the data layout (address space size, I think you could have something like a 128 bit address space or just a 16 bit one).

david-arm replied:
Again, I put this code because in order to reuse `getStartAndEndForAccess` I have to pass in a SCEV value for the backedge taken count, since the original version uses it. That also means I can no longer call `getSmallConstantMaxTripCount` since that returns an unsigned value, however `getSmallConstantMaxTripCount` returns 0 if the max count > 32 bits. In order to maintain the same behaviour I added the explicit check here. I'm happy to remove it though if you feel comfortable with removing this restriction?

david-arm committed:
Address review comments

fhahn commented on llvm/lib/Analysis/Loads.cpp:
```diff
@@ -276,84 +277,80 @@ static bool AreEquivalentAddressValues(const Value *A, const Value *B) {
-
-  // If given a uniform (i.e. non-varying) address, see if we can prove the
-  // access is safe within the loop w/o needing predication.
-  if (L->isLoopInvariant(Ptr))
```
It looks like this code path was previously not tested and dropping it may cause a regression. I added 747f7f38bddaa4ae60c8d1380c9b28d81fc13e8a which should cover this codepath.

david-arm replied:
No problem. Thanks for adding the test! I've re-added the code now.

david-arm committed:
Re-add code for invariant pointers

fhahn reviewed (approved):
LGTM, thanks!

Just one comment remaining re moving the existing comment for getStartAndEndForAccess

david-arm committed:
Address review comment

david-arm commented:
Rebase + addressed review comment. I ran "make check-all" downstream after rebasing and the tests look good.

david-arm commented:
Looks like this is causing an assert failure for build https://lab.llvm.org/buildbot/#/builders/169/builds/7413. I'll revert asap.
